### PR TITLE
add log-sum-exp "trick" for numerical stability

### PR DESCRIPTION
**Pull request by JohnGiorgi**
This PR updates the `n_pair_loss` method so that it uses the [log-sum-exp](https://en.wikipedia.org/wiki/LogSumExp) trick for numerical stability. When I used the `NPairLoss` within my model, I ran into NaN's.

It would be better to use [PyTorch's implementation](https://pytorch.org/docs/stable/torch.html#torch.logsumexp) directly, but I wasn't sure what to do with the 1 + within the log of the loss function, so I rolled my own solution. I tested that it works in my own model.

### CODE REVIEW

**Comment by leeesangwon** [comment]
- You may find the meaning of the `log(1+x)` at [the original N-pair loss paper](http://www.nec-labs.com/uploads/images/Department-Images/MediaAnalytics/papers/nips16_npairmetriclearning.pdf) (equation 6) equation 6

- The original equation and your modified equation seems to have a little difference.
  ```
  log(1+sum(exp(x)))  
  = log(1+sum(exp(x-a+a))) 
  = log(1+exp(a)*sum(exp(x-a))) 
  = log(exp(a)*(1/exp(a) + sum(exp(x-a)))) 
  = a + log(exp(-a)+sum(exp(x-a))
  != a + log(1+sum(exp(x-a)))  # your modified version
  ```
  Even though the difference looks tiny and you reported your modification had worked well, because “N-pair loss” is not my idea, I hope to preserve the equation as written in the paper.
  Therefore I think the line 144 of your PR should become like below.
  ```
  loss = torch.mean(a + torch.log(torch.exp(-a) + x))
  ```

- However for now I don’t have enough computation resource and time so it is hard to test the new one.

- If you have additional comments, please do not hesitate contact me.

- Thank you for your feedback!
